### PR TITLE
dissect: add a mode for operating on an in-memory copy of a DDI, inst…

### PR DESCRIPTION
**Pull request by poettering**
…ead of directly on it

This is useful for operating in ephemeral, writable mode on any image, including read-only ones. It also has the benefit of not keeping the image file's filesystem busy.

Inspired by the discussions in #25648

### CODE REVIEW

File: src/shared/data-fd-util.c [hunk 1]
```diff
@@ -343,3 +348,49 @@ int copy_data_fd(int fd) {
...
+        int r;
+
+        /* Creates a clone of a regular file in a memfd. Unlike copy_data_fd() this returns strictly a memfd
+         * (and if it can't it will fail). Thus the resulting fd is seekable, and definitely reports as
+         * S_ISREG. */
+
+        assert(fd >= 0);
+        assert(IN_SET(mode & O_ACCMODE, O_RDONLY, O_RDWR));
+        assert((mode & ~(O_RDONLY|O_RDWR|O_CLOEXEC)) == 0);
+
```
Comment (bluca) [inline src/shared/data-fd-util.c:365]: `assert(name)`

File: src/dissect/dissect.c [hunk 2]
```diff
@@ -106,6 +107,7 @@ static int help(void) {
                "     --mkdir              Make mount directory before mounting, if missing\n"
                "     --rmdir              Remove mount directory after unmounting\n"
                "     --discard=MODE       Choose 'discard' mode (disabled, loop, all, crypto)\n"
+               "     --memory             Copy image into memory\n"
```
Comment (bluca) [inline src/dissect/dissect.c]: A bit generic, maybe --from-memory or --in-memory? As in "dissect mount in memory"

**Comment by poettering** [comment]
force pushed new version, addressing @bluca's points

**Review by bluca — APPROVED** [review]

File: man/systemd-dissect.xml [hunk 3]
```diff
@@ -309,6 +309,15 @@
         systems. If <literal>all</literal> discarding is unconditionally enabled.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--in-memory</option></term>
+
+        <listitem><para>If specified an in-memory copy of the specified disk image is used. This may be used
+        to operate with write-access on a (possibly read-only) image, without this actually modifying the
+        original file. This may also be used in order to operate on a disk image without keeping the
```
Comment (keszybz) [inline man/systemd-dissect.xml]: without this actually → without actually

File: src/shared/data-fd-util.c [hunk 4]
```diff
@@ -343,3 +348,50 @@ int copy_data_fd(int fd) {
...
+                rfd = fd_reopen(mfd, mode);
+                if (rfd < 0)
+                        return rfd;
+
+                return TAKE_FD(rfd);
+        }
+
+        off_t f = lseek(mfd, 0, SEEK_SET);
+        if (f != 0)
+                return -errno;
```
Comment (keszybz) [inline src/shared/data-fd-util.c]:
```
if (lseek(mfd, 0, SEEK_SET) < 0)
    return -errno;
```
?

File: src/shared/data-fd-util.c [hunk 5]
```diff
@@ -343,3 +348,50 @@ int copy_data_fd(int fd) {
...
+
+        ro = (mode & O_ACCMODE) == O_RDONLY;
+
+        mfd = memfd_create(name,
+                           ((FLAGS_SET(mode, O_CLOEXEC) || ro) ? MFD_CLOEXEC : 0) |
+                           (ro ? MFD_ALLOW_SEALING : 0));
+        if (mfd < 0)
+                return -errno;
+
+        r = copy_bytes(fd, mfd, UINT64_MAX, COPY_REFLINK);
```
Comment (keszybz) [inline src/shared/data-fd-util.c:374]: COPY_REFLINK can be unset here. We know it cannot work.

File: src/shared/data-fd-util.c [hunk 6]
```diff
@@ -343,3 +348,50 @@ int copy_data_fd(int fd) {
 
         return fd_reopen(tmp_fd, O_RDONLY|O_CLOEXEC);
 }
+
+int memfd_clone_fd(int fd, const char *name, int mode) {
+        _cleanup_close_ int mfd = -EBADF;
```
Comment (keszybz) [inline src/shared/data-fd-util.c:353]: It doesn't matter greatly, but we always use -1 for fd initialization. Even in this function, -1 is used below.

File: src/shared/loop-util.c [hunk 7]
```diff
@@ -640,6 +641,47 @@ int loop_device_make_by_path(
...
 }
 
+int loop_device_make_by_path_memory(
+                const char *path,
+                int open_flags,
+                uint32_t loop_flags,
+                int lock_op,
+                LoopDevice **ret) {
+
+        _cleanup_close_ int fd = -EBADF, mfd = -EBADF;
```
Comment (keszybz) [inline src/shared/loop-util.c:651]: Same here.

File: src/shared/data-fd-util.h [hunk 8]
```diff
@@ -5,3 +5,4 @@
 
 int acquire_data_fd(const void *data, size_t size, unsigned flags);
 int copy_data_fd(int fd);
+int memfd_clone_fd(int fd, const char *name, int mode);
```
Comment (keszybz) [inline src/shared/data-fd-util.h:8]: I'd prefer the addition of memfd_clone_fd() to be a separate commit, even if it doesn't have a test. The function is large enough to deserve to be its own thing.

**Review by keszybz — COMMENTED** [review]
LGTM too. Some minor suggestions, but nothing I'd insist on.

File: src/shared/data-fd-util.c (hunk 5 above)
Comment (poettering) [reply src/shared/data-fd-util.c:374]: that's not that obvious. First of all, memfds are all attached to the same superblock, so reflinks might actually work if shmemfs would one day support it – in case input fd is already a memfd. There's not much lost for trying.

Also I think one day the same-superblock limitations of reflinks might get removed too, because btrfs for example exposes multiple superblocks on the same fs (each subvol is one), and they are working on allowing reflinks between subvols iirc.

Hence, when that materializes then sooner or later reflinks between tmpfs/shmfs/memfd superblocks might end up being permitted too...

So I wouldn't make unnecessary restrictions here. it's not that unlikely that this works eventually.

File: src/shared/data-fd-util.c (hunk 6 above)
Comment (poettering) [reply src/shared/data-fd-util.c:353]: oh, see discussion on #25513. LXC codebases do it like this, and I think it makes sense to just say -EBADF is the "invalid" value for fds, rather than the usual -1?

So we started merging code with -EBADF instead of -1 as "bad fd" indicator, and I think it's a scheme that makes a ton of sense and we should follow for new code everywhere.

**Comment by poettering** [comment]
force pushed new version, addressing @keszybz' points, except as discussed above
